Sources/Workspace/Workspace.swift

```diff
@@ -156,22 +102,46 @@ private class WorkspaceRepositoryManagerDelegate: RepositoryManagerDelegate {
         workspaceDelegate.fetchingWillBegin(repository: handle.repository.url, fetchDetails: details)
     }
 
-    func fetchingDidFinish(handle: RepositoryManager.RepositoryHandle, fetchDetails details: RepositoryManager.FetchDetails?, error: Swift.Error?) {
+    func fetchingDidFinish(handle: RepositoryManager.RepositoryHandle, fetchDetails details: RepositoryManager.FetchDetails?, error: Swift.Error?, duration: DispatchTimeInterval) {
         let diagnostic: Diagnostic? = error.flatMap({
             let engine = DiagnosticsEngine()
             engine.emit($0)
             return engine.diagnostics.first
         })
-        workspaceDelegate.fetchingDidFinish(repository: handle.repository.url, fetchDetails: details, diagnostic: diagnostic)
+        workspaceDelegate.fetchingDidFinish(repository: handle.repository.url, fetchDetails: details, diagnostic: diagnostic, duration: duration)
     }
 
     func handleWillUpdate(handle: RepositoryManager.RepositoryHandle) {
         workspaceDelegate.repositoryWillUpdate(handle.repository.url)
     }
 
-    func handleDidUpdate(handle: RepositoryManager.RepositoryHandle) {
-        workspaceDelegate.repositoryDidUpdate(handle.repository.url)
+    func handleDidUpdate(handle: RepositoryManager.RepositoryHandle, duration: DispatchTimeInterval) {
+        workspaceDelegate.repositoryDidUpdate(handle.repository.url, duration: duration)
+    }
+}
+
+private struct WorkspaceDependencyResolverDelegate: DependencyResolverDelegate {
+    unowned let workspaceDelegate: WorkspaceDelegate
+
+    init(_ delegate: WorkspaceDelegate) {
+        self.workspaceDelegate = delegate
+    }
+
+    func willResolve(term: Term) {
+        self.workspaceDelegate.willComputeVersion(package: term.node.package.identity, location: term.node.package.location)
+    }
+
+    func didResolve(term: Term, version: Version, duration: DispatchTimeInterval) {
+        self.workspaceDelegate.didComputeVersion(package: term.node.package.identity, location: term.node.package.location, version: version.description, duration: duration)
```

Comment:
This is particularly great — is it guaranteed that the package is "done" at this point, i.e. checked out to the right version and there is no chance that PubGrub might go back and check out a different version?